util/types/src/core/extras.rs

```diff
@@ -231,3 +243,71 @@ impl EpochExtBuilder {
         self.0
     }
 }
+
+#[derive(Debug, Default, Clone, Eq, PartialEq, Hash)]
+pub struct DetailedEpochNumber(u64);
+
+impl fmt::Display for DetailedEpochNumber {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl FromStr for DetailedEpochNumber {
+    type Err = ParseIntError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let v = u64::from_str(s)?;
+        Ok(DetailedEpochNumber(v))
+    }
+}
+
+impl DetailedEpochNumber {
+    pub const NUMBER_BITS: usize = 32;
+    pub const NUMBER_MAXIMUM_VALUE: u64 = (1u64 << Self::NUMBER_BITS);
+    pub const NUMBER_MASK: u64 = (Self::NUMBER_MAXIMUM_VALUE - 1);
+    pub const FRACTION_BITS: usize = 16;
+    pub const FRACTION_MAXIMUM_VALUE: u64 = (1u64 << Self::FRACTION_BITS);
+    pub const FRACTION_MASK: u64 = (Self::FRACTION_MAXIMUM_VALUE - 1);
+    pub const NUMBER_WITH_FRACTION_BITS: usize = Self::NUMBER_BITS + Self::FRACTION_BITS;
+    pub const NUMBER_WITH_FRACTION_MAXIMUM_VALUE: u64 = (1u64 << Self::NUMBER_WITH_FRACTION_BITS);
+    pub const NUMBER_WITH_FRACTION_MASK: u64 = (Self::NUMBER_WITH_FRACTION_MAXIMUM_VALUE - 1);
+    pub const LENGTH_BITS: usize = 16;
+    pub const LENGTH_MAXIMUM_VALUE: u64 = (1u64 << Self::LENGTH_BITS);
+    pub const LENGTH_MASK: u64 = (Self::LENGTH_MAXIMUM_VALUE - 1);
+
+    pub fn new(number: u64, fraction: u64, length: u64) -> DetailedEpochNumber {
+        debug_assert!(number < Self::NUMBER_MAXIMUM_VALUE);
+        debug_assert!(fraction < Self::FRACTION_MAXIMUM_VALUE);
+        debug_assert!(length < Self::LENGTH_MAXIMUM_VALUE);
+        DetailedEpochNumber(
+            (length << Self::NUMBER_WITH_FRACTION_BITS)
+                | (number << Self::FRACTION_BITS)
+                | fraction,
+        )
+    }
+
+    pub fn number(&self) -> u64 {
```

Comment:
The `DetailedEpochNumber ` and `EpochNumber` both exist in the current implementation, are we still use `EpochNumber` to indicate the `number` part? if not the `primitive ::EpochNumber` alias should be removed.